src/Libraries/RevitNodes/GeometryConversion/CurveLoopPartition.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Autodesk.DesignScript.Runtime;
+using Autodesk.Revit.DB;
+using BoundingBox = Autodesk.DesignScript.Geometry.BoundingBox;
+using PolyCurve = Autodesk.DesignScript.Geometry.PolyCurve;
+
+namespace Revit.GeometryConversion
+{
+
+    [SupressImportIntoVM]
+    internal static class CurveLoopPartition
+    {
+        public static List<List<List<Curve>>> ByCurveLoopsAndFace(Face face, List<List<Curve>> curveLoops)
+        {
+            if (curveLoops.Count == 1)
+                return new List<List<List<Curve>>>()
+                {
+                    curveLoops
+                };
+
+            var tesselatedCurveLoops = new List<Tuple<BoundingBox, List<XYZ>, List<Curve>>>();
+            foreach (var curveLoop in curveLoops)
+            {
+                var verts = TesselateCurveLoop(curveLoop).ToList();
+                List<Autodesk.DesignScript.Geometry.Curve> curves =
+                    new List<Autodesk.DesignScript.Geometry.Curve>();
+                foreach (var curve in curveLoop)
+                {
+                    curves.Add(curve.ToProtoType(false));
+                }
+                var geom = new List<Autodesk.DesignScript.Geometry.Geometry>();
+                geom.Add(PolyCurve.ByJoinedCurves(curves));
+                tesselatedCurveLoops.Add(new Tuple<BoundingBox, List<XYZ>, List<Curve>>(BoundingBox.ByGeometry(geom), verts, curveLoop));
+                curves.ForEach(x => x.Dispose());
+                curves.Clear();
+                geom.ForEach(x => x.Dispose());
+                geom.Clear();
+            }
+
+            var outerCurveLoops = new List<Tuple<BoundingBox, List<XYZ>, List<Curve>>>();
+            var innerCurveLoops = new List<Tuple<BoundingBox, List<XYZ>, List<Curve>>>();
+
+            // collect the outer loops, which are not contained in any other loop
+            for (int i = 0, j = tesselatedCurveLoops.Count - 1; i < tesselatedCurveLoops.Count; j = i++) 
+            {
+                var curveLoop = tesselatedCurveLoops[i];
+                var isOuter = tesselatedCurveLoops.Where((_, l) => i != l)
+                    .All(bound => !curveLoop.Item2.IsContainedIn(bound.Item1));
+
+                if(isOuter)
+                {
+                    outerCurveLoops.Add(curveLoop);
+                }
+                else
+                {
+                    innerCurveLoops.Add(curveLoop);
+                }
+            }
+
+            var partitionedLoops = new List<List<List<Curve>>>();
+
+            // for each outer loop, collect the loops that are inside of it
+            // forming the partitions of the original curve loop set
+            foreach (var outerLoop in outerCurveLoops)
```

Comment:
Do you means that there will be several outer loops for one face?

Reply:
Yes, sometimes there would be 2 or more outer loops.